File: aula-01/js/CPF.js
/**
 * Funções para verificar validade de CPFs
 */

module.exports = {
    f1: converteCaracteresEmInteiros,
    f2: validarDigitosCPF,
    f3: validarDigitosCPF2
};

/**
 * Transforma caracteres numéricos em inteiros
 *
 * @param {String} cpf String que contém CPF
 *
 * @returns {Array} Vetor contendo em cada posição um dígito do CPF como inteiro
 */
function converteCaracteresEmInteiros(cpf) {
    var digitosCpf = [];
    for (var i = 0; i < cpf.length; i++) {
        digitosCpf[i] = parseInt(cpf.charAt(i));
    }

    return digitosCpf;
}

/**
 * Avalia se CPF fornecido é válido dado seus dígitos verificados (dois últimos dígitos)
 *
 * @param {String} cpf String que contém CPF
 *
 * @returns {boolean} Verdadeiro, se dígitos verificadores do CPF estiverem corretos, Falso, se não
 *
 * @throws {RangeError} Se cpf tem mais que 11 caracteres
 * @throws {TypeError} Se cpf é null ou undefined
 */
function validarDigitosCPF(cpf) {
    if (cpf === null || cpf === undefined){
        throw new TypeError("Argumento null ou undefined");
    }
    if (cpf.length != 11) {
        throw new RangeError("CPF deve ter 11 dígitos: " + cpf);
    }

    var digitosCpf = converteCaracteresEmInteiros(cpf);
    var calculoParcialDigito10 = digitosCpf[0];
    var calculoParcialDigito11 = digitosCpf[1];
    ;
    for (var i = 1; i < 9; i++) {
        calculoParcialDigito10 += digitosCpf[i] * (i + 1);
    }

    for (var i = 2; i < 10; i++) {
        calculoParcialDigito11 += digitosCpf[i] * i;
    }

    var calculoFinalDigito10 = (calculoParcialDigito10 % 11) % 10;
    var calculoFinalDigito11 = (calculoParcialDigito11 % 11) % 10;

    return calculoFinalDigito10 == digitosCpf[9] && calculoFinalDigito11 == digitosCpf[10];
}

/**
 * Avalia se CPF fornecido é válido dado seus dígitos verificados (dois últimos dígitos)
 *
 * @param {String} cpf String que contém CPF
 *
 * @returns {boolean} Verdadeiro, se dígitos verificadores do CPF estiverem corretos, Falso, se não
 *
 * @throws {RangeError} Se cpf tem mais que 11 dígitos
 * @throws {TypeError} Se cpf é null ou undefined
 */
function validarDigitosCPF2(cpf) {
    if (cpf === null || cpf === undefined){
        throw new TypeError("Argumento null ou undefined");
    }
    if (cpf.length != 11) {
        throw new RangeError("CPF deve ter 11 dígitos: " + cpf);
    }

    var digitosCpf = converteCaracteresEmInteiros(cpf);
    var calculoParcialDigito11 = digitosCpf[8];
    var calculoParcialDigito10 = calculoParcialDigito11;
    for (var c = 7; c >= 0; c--) {
        calculoParcialDigito11 += digitosCpf[c];
        calculoParcialDigito10 += calculoParcialDigito11;
    }
    var calculoFinalDigito10 = (calculoParcialDigito10 % 11) % 10;
    var calculoFinalDigito11 = ((calculoParcialDigito10 - calculoParcialDigito11 + 9 * digitosCpf[9]) % 11) % 10;

    return calculoFinalDigito10 === digitosCpf[9] && calculoFinalDigito11 === digitosCpf[10];
}

File: aula-01/java/CPF.java
import java.util.stream.Stream;

public class CPF {

    private String cpf;

    public CPF(String cpf) {
        validarCPF(cpf);
        this.cpf = cpf;
    }

    public String getCpf() {
        return this.cpf;
    }

    public void validarCPF(String cpf) {
        if (cpf == null) {
            throw new IllegalArgumentException("argumento é null");
        }

        if (cpf.length() != 11) {
            throw new IllegalArgumentException("CPF deve ter 11 dígitos: " + cpf);
        }

        if (!sequenciaContemApenasDigitos(cpf)) {
            throw new IllegalArgumentException("CPF deve conter somente dígitos (0 a 9): " + cpf);
        }

        if (!validarDigitosCPF2(cpf)) {
            throw new IllegalArgumentException("dígitos verificadores do CPF incorretos: " + cpf);
        }
    }

    public boolean sequenciaContemApenasDigitos(final String cpf) {
        return cpf.chars().allMatch(c -> Character.isDigit(c));
    }

    public int[] converteCaracteresEmInteiros(String cpf) {
        int[] cpfVetor = new int[11];
        for (int i = 0; i < cpf.length(); i++) {
            cpfVetor[i] = Character.getNumericValue(cpf.charAt(i));
        }
        return cpfVetor;
    }

    public boolean validarDigitosCPF(String cpf) {
        int[] cpfVetor = converteCaracteresEmInteiros(cpf);
        int j = cpfVetor[0];
        int k = cpfVetor[1];
        for (int i = 1; i < 9; i++) {
            j += cpfVetor[i] * (i + 1);
        }
        for (int i = 2; i < 10; i++) {
            k += cpfVetor[i] * i;
        }
        int dj = (j % 11) % 10;
        int dk = (k % 11) % 10;
        return dj == cpfVetor[9] & dk == cpfVetor[10];
    }

    public boolean validarDigitosCPF2(String cpf) {
        int[] cpfVetor = converteCaracteresEmInteiros(cpf);
        int p = cpfVetor[8];
        int s = p;
        for (int c = 7; c >= 0; c--) {
            p += cpfVetor[c];
            s += p;
        }
        int j = (s % 11) % 10;
        int k = ((s - p + 9 * cpfVetor[9]) % 11) % 10;
        return (j == cpfVetor[9] & k == cpfVetor[10]);
    }

}
